clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -191,8 +191,7 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   if (const FunctionDecl *FD = dyn_cast<FunctionDecl>(D)) {
     const IdentifierInfo *II = FD->getIdentifier();
     const DeclContext *DC = FD->getDeclContext();
-    if (II && II->isStr("__spirv_ocl_printf") &&
-        !FD->isDefined() &&
+    if (II && II->isStr("__spirv_ocl_printf") && !FD->isDefined() &&
```

Comment:
1. The attached file is already in Markdown format, although I was unable to attach it as a .md file. Hence it was saved as a .txt file and attached.
2. The suggestion to add the design document to sycl/doc does not seem appropriate. The document is intended to explain this patch, the implementation of this missing feature. It is not intended to be read by any user of SYCL.
2. The word "extension" in the design document referred to an extension of the implementation. Although section 4.8.10 of the SYCL spec does not explicitly mention arrays, I would argue that arrays as implemented in this PR are already covered by the spec, thus this is not an extension to the SYCL spec.
3. The one change unrelated to my patch was caused by clang-format. If you feel strongly about that one-line formatting change, I will revert it.
4. I am not sure what you mean by "CodeGen test checking resulting IR module". The test clang/test/SemaSYCL/array-kernel-param-neg.cpp is a negative test, i.e., for producing expected diagnostics. The test LLVM IR test clang/test/SemaSYCL/array-kernel-param.cpp is a positive test to check expected LLVM IR test produced by clang, and the test sycl/test/array_param/array-kernel-param-run.cpp is a runnable test. Is there another type of test you would like to see added?

Reply:
> The suggestion to add the design document to sycl/doc does not seem appropriate. The document is intended to explain this patch, the implementation of this missing feature. It is not intended to be read by any user of SYCL.

I think we need to save information describes this patch somehow in the repo, because it will be really helpful to understand what is happening here in the future. If you are not going to commit the doc (but, I don't see the problem here, we have docs which describe implementation details, for example this one https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md), please add some information to commit messages/code comments etc. 

> I am not sure what you mean by "CodeGen test checking resulting IR module". The test clang/test/SemaSYCL/array-kernel-param-neg.cpp is a negative test, i.e., for producing expected diagnostics. The test LLVM IR test clang/test/SemaSYCL/array-kernel-param.cpp is a positive test to check expected LLVM IR test produced by clang, and the test sycl/test/array_param/array-kernel-param-run.cpp is a runnable test. Is there another type of test you would like to see added?

`clang/test/SemaSYCL/array-kernel-param.cpp` which you added checks not LLVM IR, but clang AST nodes. I suggest adding test like this https://github.com/intel/llvm/blob/sycl/clang/test/CodeGenSYCL/basic-kernel-wrapper.cpp one.
AST is some clang internal representation, whereas result of clang run is LLVM IR module, which is checked in CodeGen tests.
BTW, you can also add a test checking integration header emission since you have changes there (like this one https://github.com/intel/llvm/blob/sycl/clang/test/CodeGenSYCL/wrapped-accessor.cpp )

Reply:
Just my two cents, so feel free to ignore it. However, I think the addition of the document to sycl/doc (or some variation) of it would be quite useful. Whilst it does focus on your patch the pseudo-code does help showcase some of the concepts behind SemaSYCL.cpp quite well. And considering this will all hopefully eventually be committed upstream to Clang/LLVM, it may be useful to help other future maintainers understand it! 

I might just be biased though as I find the compiler documentation quite useful and I quite liked your examples.

Reply:
Tests have been added to the test/CodeGenSYCL directory.
The document describing this change has been added to sycl/doc directory.
The unrelated clang-formatting change has been reverted.